WalletWasabi/Wallets/FilterProcessor/BlockFilterIterator.cs

```diff
@@ -0,0 +1,79 @@
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Backend.Models;
+using WalletWasabi.Stores;
+
+namespace WalletWasabi.Wallets.FilterProcessor;
+
+/// <summary>
+/// Iterator that caches block filters from the index store to allow efficient iteration over block filters in ascending order.
+/// </summary>
+public class BlockFilterIterator
+{
+	public BlockFilterIterator(IIndexStore indexStore, int maxNumberFiltersInMemory = 1000)
+	{
+		IndexStore = indexStore;
+		MaxNumberFiltersInMemory = maxNumberFiltersInMemory;
+	}
+
+	/// <remarks>Internal only to allow modifications in tests.</remarks>
+	internal Dictionary<uint, FilterModel> Cache { get; } = new();
+	private IIndexStore IndexStore { get; }
+	private int MaxNumberFiltersInMemory { get; }
+
+	/// <summary>
+	/// Gets block filter for the block of specified height.
+	/// </summary>
+	/// <remarks>Filter is immediately removed from the cache once the method returns. Repeated calls for single height are thus expensive.</remarks>
+	public async Task<FilterModel> GetAndRemoveAsync(uint height, CancellationToken cancellationToken)
+	{
+		// Each block filter is to needed just once, so we can remove the block right now and free the memory sooner.
+		if (!Cache.Remove(height, out FilterModel? result))
+		{
+			// We don't have the next filter to process, so fetch another batch of filters from the database.
+			Clear();
+
+			FilterModel[] filtersBatch = await IndexStore.FetchBatchAsync(height, MaxNumberFiltersInMemory, cancellationToken).ConfigureAwait(false);
+
+			// Check that we get a block filter and that the filter is actually the one we want as the previous command does not guarantee that we get such block.
+			if (filtersBatch.Length == 0)
+			{
+				throw new UnreachableException($"No block was found for a batch starting with block height {height}.");
+			}
+			else if (filtersBatch[0].Header.Height != height)
+			{
+				throw new UnreachableException($"Block filter for height {height} was not found.");
+			}
+
+			// Cache filters.
+			uint i = height;
+			foreach (FilterModel filter in filtersBatch)
+			{
+				// Make sure that the sequence of blocks is consecutive.
+				if (i != filter.Header.Height)
```

Comment:
If you implement my suggestion just above, it might be clearer to ditch `i` and instead compare `lastHeight + 1 == filter.Header.Height`. Again, feel free to ignore